### PR TITLE
[UIE-74] Add types for Select component

### PR DESCRIPTION
**Pull request by nawatts**
Stacked on #3804.

Motivated by the discovery that several components misuse Select (by passing a `disabled` prop instead of `isDisabled`), this PR adds type definitions for the basic Select components.

Terra UI's Select components wrap [react-select](https://www.npmjs.com/package/react-select)'s Select component. However, the types are not simple passthroughs.

Complicating the types, Terra UI adds some convenience in how options and value may be passed. For example:
```js
h(Select, {
  options: ['foo', 'bar'],
  value: 'foo'
})
```
instead of:
```js
const options = [{ value: 'foo' }, { value: 'bar' }]
h(Select, {
  options,
  value: options[0[
})
```

Also, react-select uses the same component for grouped / un-grouped options, while Terra UI separates them.

### CODE REVIEW

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts [hunk 1]
```diff
@@ -37,6 +37,8 @@ const titleId = 'analysis-modal-title'
 const analysisMode = Symbol('artifact')
 const environmentMode = Symbol('environment')
 
+const StringSelect = Select as typeof Select<string>
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts:46]: Ideally, we'd have more specific types, so "StringSelect" and the like wouldn't be a frequent appearance. The other two files in this PR have what I imagine will be more typical examples.

File: src/pages/library/SearchAndFilterComponent.ts [hunk 2]
```diff
@@ -483,13 +485,13 @@ export const SearchAndFilterComponent = <ListItem>({
...
           }, ['Sort by']),
-          h(Select, {
+          h(SortSelect, {
             id,
             isClearable: false,
             isSearchable: false,
             styles: { container: old => ({ ...old, flex: '0 0 content' }) },
             value: sort,
-            onChange: ({ value }) => setSort(value),
+            onChange: opt => setSort(opt!.value),
```
Comment (nawatts) [inline src/pages/library/SearchAndFilterComponent.ts:494]: react-select's types say that onChange may be called with null. That seemed reasonable to leave in at the type level, so I've changed code that assumed a non-null value in onChange to use non-null assertions.

File: src/pages/library/data-catalog/CreateDataset/CustomInputs/StorageInput.ts [hunk 3]
```diff
@@ -34,18 +36,18 @@ export const StorageInput = ({ wrapperProps, storageObject, onChange }: StorageI
...
-      h(SelectInput, {
+      h(CloudPlatformStorageInput, {
         title: 'Cloud Platform',
         wrapperProps: inputWrapperProps,
         value: storageObject.cloudPlatform,
         options: ['gcp', 'azure'],
-        onChange: option => onChange({ cloudPlatform: option.value })
+        onChange: option => onChange({ cloudPlatform: option!.value })
       }),
       h(SelectInput, generateSelectInputProps('Cloud Resource', 'cloudResource', azureCloudResourceTypes, googleCloudResourceTypes)),
```
Comment (nawatts) [inline src/pages/library/data-catalog/CreateDataset/CustomInputs/StorageInput.ts:52]: Still some work that could be done here to type these inputs.

**Review by marctalbott — APPROVED** [review]
Reviewed and tested out Workspaces things and all looks good

**Review by s-rubenstein — COMMENTED** [review]
Just to confirm, do you mean to merge this into `upgrade-babel`? Or should this go into dev?

**Review by s-rubenstein — APPROVED** [review]
This looks good to me. I haven't tested things out locally, but I can if we want.

**Comment by nawatts** [comment]
> Just to confirm, do you mean to merge this into upgrade-babel? Or should this go into dev?

I would like to merge #3804 to dev, then merge this to dev. For now, until #3804 is merged, this PR uses `upgrade-babel` as the base branch to isolate the changes to Select for review.

**Review by n8landolt — APPROVED** [review]

**Comment by nawatts** [comment]
Rebased on latest #3804.

**Review by blakery — APPROVED** [review]

**Comment by nawatts** [comment]
Rebased after merging #3804.

**Comment by nawatts** [comment]
Had to make a few fixups for recently added Selects. That seems to happen every time I update this branch, so I'm going to go ahead and merge this to avoid falling behind again.

**Pull request merged**